accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,500 @@
+
+NuGet should search for files whose names begin with `nuget-plugin-*` by scanning all the directories in the `PATH` environment variable.
+To ensure compatibility across different platforms, the implementation could convert all file names to lowercase before checking for a file.
+
+- On Windows, NuGet should search for plugins using the `.exe` extension.
+The `PATHEXT` environment variable in Windows specifies the file extensions that the operating system considers to be executable.
+When you enter a command without specifying an extension, Windows will look for files with the extensions listed in `PATHEXT` in the directories specified by the `PATH` environment variable.
+For example, if `PATHEXT` is set to `.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC`, and you enter the command `myprogram`, Windows will search for `myprogram.com`, `myprogram.exe`, `myprogram.bat`, and so on, in that order, in the directories listed in your `PATH` environment variable.
+It will execute the first match it finds.
+Given that .NET Tools are console applications, they should have the `.exe` extension on Windows to be considered valid plugins if the naming convention is followed.
+
+- Similarly, on other platforms, NuGet should search for plugins with the executable bit set to identify them as valid plugins.
+This is because on Mac and Linux, executable files typically don't have extensions.
+Given that .NET Tools, when installed on Linux and Mac, have a native shim with an executable bit set, they should be considered valid plugins if the naming convention is followed.
+
+Currently, `NuGet.exe` [scans all directories in the `PATH` environment variable](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs#L708-L736) to find `MSBuild.exe`.
+
+#### Plugin execution
+
+In the [current implementation](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs#L155-L181), NuGet launches the `.exe` file in a separate process when running on the .NET Framework.
+On the other hand, when running on .NET, NuGet executes the `dotnet plugin-in.dll` command in a separate process.
+
+The plan is to slightly adjust this implementation. NuGet will launch the `.exe` file in a separate process when running on Windows, irrespective of whether the runtime is .NET Framework or .NET.
+For non-Windows platforms, if the file does not have an extension, it will launch the executable directly in a separate process.
+If the file has a `.dll` extension, it will continue to execute the `dotnet nuget-plugin-name.dll` command in a separate process.
```

Comment:
A plugin with a `.dll` extension will only be executed using the dotnet CLI if the file is located under the NuGet plugin-specific `netcore` subdirectory.
This is to support backward compatibility with the NuGet plugins discovered from predefined plugin directories.
During PATH scanning, NuGet searches only for plugins with `.exe` or `.bat` extensions on Windows, and for files with the executable bit set on Linux/Mac.

I will add this comment to the spec to avoid confusion in the future.

Reply:
Updated spec to address this comment. `For non-Windows platforms, if the file does not have an extension, it will launch the executable directly in a separate process. If the file has a .dll extension, it will continue to execute the dotnet plugin-name.dll command in a separate process. It is worth noting that a plugin with a .dll extension will only be executed using the dotnet CLI if the file is located under the NuGet plugin-specific netcore subdirectory. This is to support backward compatibility with the NuGet plugins discovered from predefined plugin directories. As mentioned above, during PATH scanning, NuGet searches only for plugins with .exe or .bat extensions on Windows, and for files with the executable bit set on Linux/Mac.`